{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On this lab we will put to practice some of the concepts we have learned on this past few days.\n",
    "\n",
    "`NOTE: On this lab you should try to write all the functions yourself using only the most basic of python syntax and without functions such as len, count, sum, max, min, in, etc. Give it a try. 🧑🏻‍💻👩🏻‍💻`\n",
    "\n",
    "The cell after each exercise contains a few tests to check if your function works as expected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mod.testing import *\n",
    "import unittest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Write a function that returns the greater of two numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def greater(a,b):\n",
    "    if a > b:\n",
    "        return(a)\n",
    "    else:\n",
    "        return(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....................................................................................................\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.360s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_greater(greater)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Now write a function that returns the largest element on a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def greatest(arr):\n",
    "    sorted(arr)\n",
    "    for x in arr:\n",
    "        \n",
    "        return  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n",
    "test_greatest(greatest)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Write a function that sums all the elements on a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sum_all(arr):\n",
    "    Suma = 0\n",
    "    for n in arr:  \n",
    "        Suma += n\n",
    "    return Suma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....................................................................................................\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.324s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_sum(sum_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Write another function that multiplies all the elements on a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mult_all(arr):\n",
    "    Mult = 1\n",
    "    for n in arr:\n",
    "        Mult *= n\n",
    "    return Mult"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....................................................................................................\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.127s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_mult(mult_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Now combine those two ideas and write a function that receives a list and either \"+\" or \"*\" and outputs acordingly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def oper_all(arr, oper):\n",
    "    Suma = 0\n",
    "    Mult = 1\n",
    "    \n",
    "    for n in arr:  \n",
    "        Suma += n\n",
    "    for n in arr:\n",
    "        Mult *= n\n",
    "    \n",
    "    if oper == \"+\":\n",
    "        return Suma\n",
    "    elif oper == \"*\":\n",
    "        return Mult\n",
    "    else:\n",
    "        return(f\"shieeeeeeeeeet\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....................................................................................................\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.309s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_operations(oper_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Write a function that returns the factorial of a number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "def factorial(n):\n",
    "    if n > 1:\n",
    "        return n * factorial(n - 1)\n",
    "    else:\n",
    "        return 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....................................................................................................\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.305s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_factorial(factorial)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Write a function that takes a list and returns a list of the unique values.\n",
    "\n",
    "`NOTE: You cannot use set. 🤔`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique(arr):\n",
    "    list = []\n",
    "    for i in arr:\n",
    "        if i not in list:\n",
    "            list.append(i)\n",
    "    return list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....................................................................................................\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.596s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "# This will test your function\n",
    "test_unique(unique)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Write a function that returns the mode of a list, i.e.: the element that appears the most times.\n",
    "`NOTE: You should not use count... 🧐`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mode_counter(arr):\n",
    "    d = {}\n",
    "    count = 0\n",
    "    for i in arr:\n",
    "        if i not in d:\n",
    "            d[i]=1\n",
    "        else:\n",
    "            d[i]= d[1] + 1\n",
    "    \n",
    "    for key, value in d.items():\n",
    "        if value > count:\n",
    "            count = value\n",
    "            mode = key\n",
    "    return mode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "F.FEEEEEEEEEEEEEFFEFEE.EEFEFE.FEEFEEEEEEEF.EEEEEEEE.EEEFEEEEEEEEEEEE.EEEEEEFEFEFEEEFEEFEFE.EFEEEFEFF\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "ERROR: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "  File \"/tmp/ipykernel_6522/488659024.py\", line 8, in mode_counter\n",
      "    d[i]= d[1] + 1\n",
      "KeyError: 1\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 7 != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 5 != 9 : Should be 9\n",
      "\n",
      "======================================================================\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 11 != 3 : Should be 3\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 17 != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 21 != 6 : Should be 6\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 18 != 17 : Should be 17\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 16 != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 18 != 20 : Should be 20\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 25 != 14 : Should be 14\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 8 != 10 : Should be 10\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 24 != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 9 != 14 : Should be 14\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 7 != 17 : Should be 17\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 14 != 6 : Should be 6\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 5 != 18 : Should be 18\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 11 != 4 : Should be 4\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 20 != 10 : Should be 10\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 3 != 21 : Should be 21\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 5 != 22 : Should be 22\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 15 != 12 : Should be 12\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/c/Users/luis-/lab-functions/mod/testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: 12 != 21 : Should be 21\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.448s\n",
      "\n",
      "FAILED (failures=21, errors=72)\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_mode(mode_counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Write a function that calculates the standard deviation of a list.\n",
    "`NOTE: Do not use any libraries or already built functions. 😉`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statistics import stdev\n",
    "def st_dev(arr):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n",
    "test_stdev(st_dev)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10. Write a function to check if a string is a pangram, i.e.: if it contains all the letters of the alphabet at least once. Mind that the strings may contain characters that are not letters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pangram(string):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n",
    "test_pangram(pangram)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 11. Write a function that receives a string of comma separated words and returns a string of comma separated words sorted alphabetically.\n",
    "\n",
    "`NOTE: You may use sorted but not split and definitely no join! 🤪`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_alpha(string):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n",
    "test_alpha(sort_alpha)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12. Write a function to check if a given password is strong (at least 8 characters, at least one lower case, at least one upper case, at least one number and at least one special character). It should output True if strong and False if not.\n",
    "`Valid special characters: # @ ! $ % & ( ) ^ * [ ] { }`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_pass(string):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n",
    "test_pass(check_pass)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
